File: cc/svg.cc
#include "shade/cc/svg.h"
#include "include/core/SkStream.h"
#include "shade/src/svg.rs.h"

SvgDom::SvgDom(sk_sp<SkSVGDOM> dom) : dom_(std::move(dom)) {}

SkSize SvgDom::get_container_size() const {
  return dom_->containerSize();
}

void SvgDom::set_container_size(const SkSize& size) {
  dom_->setContainerSize(size);
}

void SvgDom::render(SkCanvas& canvas) const {
  dom_->render(&canvas);
}

std::unique_ptr<SvgDom> new_svg_dom(rust::Slice<const std::uint8_t> data) {
  if (data.size() == 0) {
    return nullptr;
  }

  SkMemoryStream stream(data.data(), data.size());
  sk_sp<SkSVGDOM> dom = SkSVGDOM::MakeFromStream(stream);
  if (!dom) {
    return nullptr;
  }

  return std::make_unique<SvgDom>(std::move(dom));
}

File: src/svg.rs
#[cxx::bridge]
mod ffi {
    unsafe extern "C++" {
        include!("shade/cc/svg.h");

        type SvgDom;

        fn new_svg_dom(data: &[u8]) -> UniquePtr<SvgDom>;

        fn get_container_size(self: &SvgDom, width: &mut f32, height: &mut f32);

        fn set_container_size(self: Pin<&mut SvgDom>, width: f32, height: f32);

        type SkCanvas = crate::core::SkCanvas;

        fn render(self: &SvgDom, canvas: Pin<&mut SkCanvas>);
    }
}

pub use self::ffi::*;

impl SvgDom {
    #[inline]
    pub fn new<T: AsRef<[u8]>>(data: T) -> cxx::UniquePtr<SvgDom> {
        new_svg_dom(data.as_ref())
    }
}
